### PR TITLE
Introduce DeleteDeclarationsRefactoringAction replacing IModuleRewritingExtensions.cs

### LINKED ISSUE #5719: Formatting error in "Property let accessor `{x}` is not used" QuickFix
**Rubberduck version information**
The info below can be copy-paste-completed from the first lines of Rubberduck's log or the About box:
```
Version 2.5.1.5788
OS: Microsoft Windows NT 10.0.18363.0, x64
Host Product: Microsoft Office 2016 x64
Host Version: 16.0.5134.1000
Host Executable: EXCEL.EXE
```
**Description**
When executing the Quick Fix for the `Property let accessor '{x}' is not used` inspection, blank lines are left in the code. Executing an `Indent Module` does remove the blank lines per indention settings.

**To Reproduce**
1. Create a `Property Let x` but don't use it
2. Run inspections
3. execute the `Remove unused declaration` | `Selected occurrence` QF

**Expected behavior**
The rewriter should call the indenter to clean up (or insert, as necessary) any extra blank lines in code. Not doing so leaves manual clean up which requires an extra parse to get the duck back into a known state.

**Screenshots**
Initial State:
2021-04-05 10_42_20-Code Inspections

Running the quick fix:
2021-04-05 10_42_34-Code Inspections

What's left behind:
2021-04-05 10_43_14-Code Inspections

After manually running the indenter:
2021-04-05 10_51_28-Microsoft Visual Basic for Applications - Satisfaction Summary Master Dev xlsm -

**Logfile**
[RubberduckLog.txt](https://github.com/rubberduck-vba/Rubberduck/files/6258768/RubberduckLog.txt)

**Additional context**
Sorry, the log file is rather large but the issue should be at/near the end.

### PR DESCRIPTION
**Pull request by BZngr**
Closes #5719

This PR introduces a `DeleteDeclarationsRefactoringAction` to replace `IModuleRewritingExtensions` - an extension class to handle deletion of Variable and Constant list declarations.  Using an extension class to handle these operations received a fairly lukewarm reception when it was introduced about a year ago.  Hopefully this refactoring action is considered a better approach.  

The PR was first started to address #5719 - but while creating the supporting refactoring action, it became clear that it was also an opportunity to consolidate all declaration deletion actions.  Consequently, the PR replaces `IModuleRewritingExtensions` and modifies `EncapsulateFieldRefactoringActions`, `MoveCloserToUsageRefactoringAction`, and the `RemoveUnusedDeclarationQuickFix` to use the `DeleteDeclarationsRefactoringAction`.

In addition to #5719, the PR also addresses 3 yet-to-be-reported issues regarding the `RemoveUnusedDeclarationQuickFix`.
1. Leaves uncompilable code when deleting a subset of Variable and Constant declarations declared in a list.  The issue is similar to #4319 (for Parameter lists) where a trailing comma is left behind in some removal scenarios.
2. Leaves uncompilable code when deleting the entire list of locally declared Constants and Variables (Leaves `Const` and `Dim` tokens behind).
3. Deleting un-used declarations leaves behind a new line character that, if there are several deletions performed, they can result in over-sized gaps between statements after the QuickFix.  These EndOfStatement newLines are now cleaned-up by the refactoring action.

To resolve the trailing comma issue, the `DeleteDeclarationsRefactoringAction` (like `IModuleRewritingExtensions`) needs to work with all the declarations-to-be-deleted as a group.  To support a QuickFix, this requirement led to adding a `FixMany(...)` procedure to the `IQuickFix` interface.  `FixMany(...)` allows a QuickFix to optionally override a default  `QuickFixBase.FixMany` implementation  in order to receive/operate on the entire `InspectionResults` set.

(Original content regarding an `Indenter` flag has been removed)

The RefactoringAction no longer uses the `Indenter`. The RefactoringAction now works directly with contexts associated with the target Declaration(s) to remove content. Changing from the `Indenter` using to `ParserRuleContext` modification was done because a RefactoringAction is intended to work along with any number of other RefactoringActions.  Invoking the `Indenter` as part of a RefactoringAction could affect content outside of the this RefactoringAction's scope and the `RewriteSession`.  This could cause trouble for other subsequent Refactoring Actions.

Other features:

- Modifies surrounding `VBAParser.EndOfStatementContexts` of retained declarations.  The modifications maintain whatever indentation and spacing scheme was present in the original code. 
- Supports the removal of `Annotations` uniquely associated with the deleted Declarations.  
- Removes `Comments` on the same logical line as the deleted Declaration.
- Supports revising `Comments` in the vicinity of a deleted declaration to include a TODO item.  It's purpose is to identify potentially invalid comments resulting from the deletion of one or more declarations. 
- Adds support for deletion of `EnumerationMember` and `Enumeration` declarations .
- Adds supports for deletion of `UserDefinedTypeMember` and `UserDefinedType` declarations

### CODE REVIEW

**Comment by BZngr** [comment]
Deleting a `Declaration` also needs to handle the `Declaration's` `AnnotationContexts` and `CommentContexts`.  This has proven to be a more interesting than expected.  Below is an approach I am currently considering and  I wanted to solicit some comments/input before proceeding very far into an implementation.  Items 1 through 4 represent potential content within the `VBAParser.EndOfStatementContext` that needs to be addressed by the deletion logic. 

`Private fizz As Long` (_begin `EndOfStatementContext`_)<1. Declaration Logical Line Comment>
<2. Comments after Declaration logical line, but preceding first Annotation>
<3. Annotation(s)>
<4. Comments within, or subsequent to, Annotations>(_end `EndOfStatementContext`_)
`Private fuzz As Long`

If `fizz` is the deletion target:
Delete 1. - assumes Comments on same logical line as a `Declaration` relates to the `Declaration`
Modify 2.  - insert "'TODO: Verify Comment is Valid: " ahead of the existing comment - assumes `CommentContexts` in this location cannot be associated reliably with the preceding or the following `Declaration`.

If `fuzz` is the deletion target:
Modify 2.  - insert "'TODO: Verify Comment is Valid: " ahead of the existing comment - assumes `CommentContexts` in this location cannot be associated reliably with the preceding or the following `Declaration`.
Delete 3. - assumes `AnnotationContexts` relate to the next `Declaration`
Delete 4. - assumes `CommentContexts` interspersed between, or subsequent to, `AnnotationContexts` relate to the next `Declaration`

**Comment by retailcoder** [comment]
Unless I'm mistaken, this annotation would *not* be recognized and associated with the declaration:

```vb
Dim thing As Long ' @SomeAnnotation
```
However this one *could*:

```vb
'@SomeAnnotation
Dim thing As Long
```

So the annotation(s) to remove would be *before* the declaration; if the declaration doesn't have items in its `Annotations` list then even if there's a preceding comment/annotation we know it doesn't need to be removed (but the idea of adding a TODO to verify whether the comment is still warranted is a good one - or perhaps UX-wise it would be best to just prompt whether to delete it or not?).

In other words you shouldn't need to work out whether there's an annotation to keep around, because the resolver already associated the annotations with the corresponding declaration.

**Comment by BZngr** [comment]
Completely agree with the comments regarding `Annotations`.  I can see how my example write-up might imply that the code would be determining `Annotation` associations - but, it will not.  The code will definitely use the resolver results to organize `Annotation` associations.


>or perhaps UX-wise it would be best to just prompt whether to delete it or not?

My expectation was that `DeleteDeclarationsRefactoringAction` would never be used in an interactive way.  I expected it to be used exclusively as a supporting refactoring action that makes `Declaration` deletions consistent across QuickFixes and other Refactorings.  The uncertainty of a comment's purpose/scope led to the "`TODO comment" injection to avoid involving the user.

Which generates another question:
 
With the exception of deleting an unused `Declaration`, this refactoring action will **always** introduce uncompilable code unless other code handles the `IdentifierReferences`.  Is there an expectation/assumption that a "refactoring action" _should always_ result in compilable code?  If so, should some other naming convention be applied to this type of supporting component?

**Comment by retailcoder** [comment]
Refactorings should always safely modify the code in a way that doesn't alter its functionality. As such, an action that leaves uncompilable code behind is somewhat incomplete and the user must manually finish the refactoring - however since Rubberduck's mind-reading capabilities are as limited as yours & mine, IMO a simple message telling the user explicitly that the refactoring is incomplete and some further manual clean-up is required is in order - now, 100% agreed, it wouldn't be the responsibility of the refactoring action itself to pop that warning - but the model should probably have a flag set that tells a calling presenter whether or not popping such a warning is warranted. That way a quick-fix or perhaps a "Safe Delete" refactoring could act accordingly while still being able to complete without user interactions.

**Comment by BZngr** [comment]
Some examples of the `Comment` and `Annotation` handling within this PR.

Starting with:
```
Option Explicit

'@Ignore VariableNotAssigned
Private bigValue As Long 'This field should never be less than 1,000,000
Private smallValue As Byte

Public Sub SetNewValue(ByVal aVal As Long)
    smallValue = aVal
End Sub
```
(Below) Current result if `bigValue` is removed by a `QuickFix` because it is flagged as unused:
```
Option Explicit

'@Ignore VariableNotUsed
'This field should never be less than 1,000,000
Private smallValue As Byte

Public Sub SetNewValue(ByVal aVal As Long)
    smallValue = aVal
End Sub
```
(Above) Note: `Annotation` is now incorrect and residual `Comment` is invalid/confusing

Result after this PR:
```
Option Explicit

Private smallValue As Byte

Public Sub SetNewValue(ByVal aVal As Long)
    smallValue = aVal
End Sub
```
(Above) Note : `Annotation` is deleted with the declaration and  `Comment` is removed because it is on the same logical line as the declaration.



And, starting from:
```
Option Explicit

'This field should never be less than 1,000,000
'@Ignore VariableNotAssigned
Private bigValue As Long
Private smallValue As Byte

Public Sub SetNewValue(ByVal aVal As Long)
    smallValue = aVal
End Sub
```
Current result if `bigValue` is removed by a `QuickFix`:
```
Option Explicit

'This field should never be less than 1,000,000
'@Ignore VariableNotAssigned
Private smallValue As Byte

Public Sub SetNewValue(ByVal aVal As Long)
    smallValue = aVal
End Sub
```
(Above) Note: Same residual `Comment` and `Annotation` problems as in the first example.


(Below) Result after this PR:
```
Option Explicit

'TODO Is this comment still valid? => This field should never be less than 1,000,000
Private smallValue As Byte

Public Sub SetNewValue(ByVal aVal As Long)
    smallValue = aVal
End Sub
```

**Comment by BZngr** [comment]
This PR should be merged after #5692 and the resulting conflicts are resolved.
